File: 0160-intersection-of-two-linked-lists/python/Solution.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def getIntersectionNode_without_len(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        p = headA
        q = headB
        while p != q:
            p = headB if not p else p.next
            q = headA if not q else q.next
        return p

    def getIntersectionNode_with_len(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        tmp1 = headA
        length1 = 0
        tmp2 = headB
        length2 = 0
        while tmp1:
            length1 += 1
            tmp1 = tmp1.next
        while tmp2:
            length2 += 1
            tmp2 = tmp2.next
        diff = length1 - length2
        longer = headA
        shorter = headB
        if diff < 0:
            diff = abs(diff)
            longer = headB
            shorter = headA
        for i in range(diff):
            longer = longer.next
        while longer != shorter:
            longer = longer.next
            shorter = shorter.next
        return longer

File: 0160-intersection-of-two-linked-lists/javascript/Solution.js

// Definition for singly-linked list.
function ListNode(val) {
  this.val = val;
  this.next = null;
}


/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function (headA, headB) {
  let lenA = 0, lenB = 0;
  for (let p = headA; p; p = p.next, ++lenA);
  for (let p = headB; p; p = p.next, ++lenB);
  let p = headA, q = headB;
  if (lenA > lenB) {
    for (let i = 0; i < lenA - lenB; ++i) p = p.next;
  } else {
    for (let i = 0; i < lenB - lenA; ++i) q = q.next;
  }
  while (p !== q) {
    p = p.next;
    q = q.next;
  }
  return p;
};

function getIntersectionNodeWithoutLen(headA, headB) {
  if (!headA || !headB) {
    return null;
  }
  let p = headA, q = headB;
  while (p !== q) {
    p = p === null ? headB : p.next;
    q = q === null ? headA : q.next;
  }
  return p;
}